File: Medium/377. Combination Sum IV/Solution.js
let dp = [];

let solve = (nums, target, val) => {
    if (val > target)
            return 0;
        if (val == target)
            return 1;

        if (dp[val] != -1)
            return dp[val];

        let ret = 0;
        for (const it of nums)
            ret += solve(nums, target, val + it);
        return dp[val] = ret;
};

var combinationSum4 = function(nums, target) {
    dp = new Array(target).fill(-1);
    return solve(nums, target, 0);
};

File: Medium/377. Combination Sum IV/Solution.cpp
class Solution {
public:

    int combinationSum4(vector<int>& nums, int target) {
        int dp[target];
        memset(dp, -1, sizeof dp);
        function<int(int)> solve = [&](int val) -> int {
            if(val > target) return 0;
            if(val == target) return 1;

            auto & ret = dp[val];
            if(~ret) return ret;

            ret = 0;
            for(auto & it : nums)
                ret += solve(val + it);
            return ret;
        };
        return solve(0);
    }
};